stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+        let txid = new_transaction.txid();
+        let new_transactions = if aggregate_key.is_some() {
+            // We do not enforce a block contain any transactions except the aggregate votes when it is NOT already set
+            info!(
+                "Signer #{}: Already has an aggregate key for reward cycle {}. Do not broadcast the transaction ({txid:?}).",
+                self.signer_id, self.reward_cycle
+            );
+            vec![]
+        } else {
+            let mut new_transactions = self.get_signer_transactions(stacks_client, current_reward_cycle).map_err(|e| {
+            warn!("Signer #{}: Failed to get old transactions: {e:?}. Potentially overwriting our existing stackerDB transactions", self.signer_id);
+        }).unwrap_or_default();
+            new_transactions.push(new_transaction);
+            new_transactions
+        };
+        let signer_message = SignerMessage::Transactions(new_transactions);
+        self.stackerdb.send_message_with_retry(signer_message)?;
+        info!(
+            "Signer #{}: Broadcasted DKG vote transaction ({txid}) to stacker DB",
+            self.signer_id,
+        );
+        Ok(())
+    }
+
+    /// Process a signature from a signing round by deserializing the signature and
+    /// broadcasting an appropriate Reject or Approval message to stackerdb
+    fn process_signature(&mut self, signature: &Signature) {
+        // Deserialize the signature result and broadcast an appropriate Reject or Approval message to stackerdb
+        let Some(aggregate_public_key) = &self.coordinator.get_aggregate_public_key() else {
+            debug!(
+                "Signer #{}: No aggregate public key set. Cannot validate signature...",
+                self.signer_id
+            );
+            return;
+        };
+        let message = self.coordinator.get_message();
+        // This jankiness is because a coordinator could have signed a rejection we need to find the underlying block hash
+        let signer_signature_hash_bytes = if message.len() > 32 {
+            &message[..32]
+        } else {
+            &message
+        };
+        let Some(signer_signature_hash) =
+            Sha512Trunc256Sum::from_bytes(signer_signature_hash_bytes)
+        else {
+            debug!("Signer #{}: Received a signature result for a signature over a non-block. Nothing to broadcast.", self.signer_id);
+            return;
+        };
+
+        // TODO: proper garbage collection...This is currently our only cleanup of blocks
+        self.blocks.remove(&signer_signature_hash);
+
+        // This signature is no longer valid. Do not broadcast it.
+        if !signature.verify(aggregate_public_key, &message) {
+            warn!("Signer #{}: Received an invalid signature result across the block. Do not broadcast it.", self.signer_id);
+            // TODO: should we reinsert it and trigger a sign round across the block again?
```

Comment:
I'm not sure how this is reachable if the signers are honest?  Is it because at a reward cycle boundary, some signers may use the current DKG key shares while others use the old ones?

Reply:
I think I should actually check against the aggrgate key approved in the contract rather than the locally stored one (will change). 

However, this path does hit if for some reason a signer triggers a new DKG round before it broadcasts the signature. Granted this shoudl NOT happen and is indicative of a bug/malformed state in the signer.